### ISSUE TITLE
GraphqlErrors do not serliaize to JSON as per spec

### ISSUE THREAD

**Issue by bbakerman** [graphql-java/graphql-java#649]
I used this test code to cause an exception while fetching in HttpMain.

```
        TracingInstrumentation instrumentation = new TracingInstrumentation() {
            @Override
            public DataFetcher<?> instrumentDataFetcher(DataFetcher<?> dataFetcher, InstrumentationFieldFetchParameters parameters) {
                String fieldName = parameters.getEnvironment().getFieldDefinition().getName();
                if (fieldName.equalsIgnoreCase("appearsIn")) {
                    return (DataFetcher<Object>) environment -> {
                        throw new RuntimeException("Bang!");
                    };
                }
                return super.instrumentDataFetcher(dataFetcher, parameters);
            }
        };

```

This causes a ErrorWhileDataFetching to be created.  In Java terms the the error object follows specification because it has a getMessage() / getPath() and so on.

It also has some graphql-java extensions such as the underlying exception (ok) and the enum classifiier via getErrorType()

However none of this matters when it gets put into JSON.  The JSON libraries (in this case GSON but I think others do the same) take the raw object and start sending its fields as they occur in the object.

Now I know the graphql-java does not want to be concerned with JSON serialization and so on but the fact is that without major rework by each graphql-java consumer, they are going to get error messages that dont follow spec.

http://facebook.github.io/graphql/#sec-Errors

The data I saw was this for a DataFetcher error : 

```
"errors": [
    {
      "path": {
        "parent": {
          "parent": {
            "parent": null,
            "segment": null,
            "pathList": []
          },
          "segment": {
            "value": "hero"
          },
          "pathList": [
            "hero"
          ]
        },
        "segment": {
          "value": "appearsIn"
        },
        "pathList": [
          "hero",
          "appearsIn"
        ]
      },
      "exception": {
        "detailMessage": "Bang!",
        "stackTrace": [
          {
            "declaringClass": "example.http.HttpMain$1",
            "methodName": "lambda$instrumentDataFetcher$0",
            "fileName": "HttpMain.java",
            "lineNumber": 105
          },
          {
            "declaringClass": "graphql.execution.ExecutionStrategy",
            "methodName": "fetchField",
            "fileName": "ExecutionStrategy.java",
            "lineNumber": 218
          },
          ... much stuff
          {
            "declaringClass": "java.lang.Thread",
            "methodName": "run",
            "fileName": "Thread.java",
            "lineNumber": 745
          }
        ],
        "suppressedExceptions": []
      },
      "sourceLocation": {
        "line": 4,
        "column": 5
      }
    }
  ]
```

Even a validation error is incorrect.  For example an invalid field comes out as

```
  "errors": [
    {
      "msg": "Validation error of type FieldUndefined: Field invalidField is undefined",
      "validationErrorType": "FieldUndefined",
      "sourceLocations": [
        {
          "line": 4,
          "column": 5
        }
      ],
      "description": "Field invalidField is undefined"
    }
  ],
```

This is wrong because the key is "msg" and "sourceLocations" and not "message" and "locations" as per spec

I propose we change the code so that graphql errors JUST WORK with a simple `frameworkOfChoice.toJson(errorList)` without them having to remap them.

We can do this perhaps by making the implementations have the right "field names" or by having them implement a "toSpecification" method that creates a map like we do in ExecutionResult.

if we do that latter, we would do that remapping if they called `executionResult.toSpecification` say.

**Comment by bbakerman** [comment]
The spec defined fields are

`message` : String
`locations` : [ { `line` : Int, `column` : Int } ]
`path` : [String or Int ]

so

    "errors": [
    {
      "message": "Name for character with ID 1002 could not be fetched.",
      "locations": [ { "line": 6, "column": 7 } ],
      "path": [ "hero", "heroFriends", 1, "name" ]
    }
  ],

We are a long way off from that when a JSON serializer is put in place

**Comment by andimarek** [comment]
I propose to follow our convention we used for `ExecutionResult` and implement a `toSpecification`: This way we don't have a implicit `don't rename private fields` rule and we would make clear what we support.

This would also solve this PR: https://github.com/graphql-java/graphql-java/pull/647 . Custom data can be added after `toSpecification`.

**Comment by bbakerman** [comment]
I have done some more digging on this by comparing Jackson and GSON and how they serilaize

Turns out for this code

        private ExecutionResult createER() {
        List<GraphQLError> errors = new ArrayList<>();

        errors.add(new ExceptionWhileDataFetching(mkPath(), new RuntimeException("Bang"), mkLocation(666, 999)));
        errors.add(new ValidationError(ValidationErrorType.UnknownType, mkLocations(), "Test ValidationError"));

        return new ExecutionResultImpl(null, errors);
    }

    private List<SourceLocation> mkLocations() {
        return stream(new SourceLocation[]{mkLocation(666, 999), mkLocation(333, 0)}).collect(toList());
    }

    private SourceLocation mkLocation(int line, int column) {
        return new SourceLocation(line, column);
    }

    private ExecutionPath mkPath() {
        return ExecutionPath.rootPath().segment("a").segment(0).segment("b").segment("c").segment(4);
    }


That Jackson works as expected :

```
{
  "data": null,
  "errors": [
    {
      "path": [
        "a",
        0,
        "b",
        "c",
        4
      ],
      "exception": {
        "cause": null,
        "stackTrace": [
          {
            "methodName": "createER",
            "fileName": "ExecutionResultTesting.java",
            "lineNumber": 59,
            "className": "example.http.ExecutionResultTesting",
            "nativeMethod": false
          },
          {
            "methodName": "testJackson",
            "fileName": "ExecutionResultTesting.java",
            "lineNumber": 47,
            "className": "example.http.ExecutionResultTesting",
            "nativeMethod": false
          },
          ...
        ],
        "message": "Bang",
        "localizedMessage": "Bang",
        "suppressed": []
      },
      "message": "Exception while fetching data (/a[0]/b/c[4]) : Bang",
      "locations": [
        {
          "line": 666,
          "column": 999
        }
      ],
      "errorType": "DataFetchingException"
    },
    {
      "validationErrorType": "UnknownType",
      "description": "Test ValidationError",
      "message": "Validation error of type UnknownType: Test ValidationError",
      "locations": [
        {
          "line": 666,
          "column": 999
        },
        {
          "line": 333,
          "column": 0
        }
      ],
      "errorType": "ValidationError"
    }
  ],
  "extensions": null
}
```

But GSON does not.  GSON only uses fields and never public getters on POJOs - boo!!

```
{
  "data": null,
  "errors": [
    {
      "path": {
        "parent": {
          "parent": {
            "parent": {
              "parent": {
                "parent": {
                  "parent": null,
                  "segment": null,
                  "pathList": []
                },
                "segment": {
                  "value": "a"
                },
                "pathList": [
                  "a"
                ]
              },
              "segment": {
                "value": 0
              },
              "pathList": [
                "a",
                0
              ]
            },
            "segment": {
              "value": "b"
            },
            "pathList": [
              "a",
              0,
              "b"
            ]
          },
          "segment": {
            "value": "c"
          },
          "pathList": [
            "a",
            0,
            "b",
            "c"
          ]
        },
        "segment": {
          "value": 4
        },
        "pathList": [
          "a",
          0,
          "b",
          "c",
          4
        ]
      },
      "exception": {
        "detailMessage": "Bang",
        "stackTrace": [],
        "suppressedExceptions": []
      },
      "sourceLocation": {
        "line": 666,
        "column": 999
      }
    },
    {
      "msg": "Validation error of type UnknownType: Test ValidationError",
      "validationErrorType": "UnknownType",
      "sourceLocations": [
        {
          "line": 666,
          "column": 999
        },
        {
          "line": 333,
          "column": 0
        }
      ],
      "description": "Test ValidationError"
    }
  ]
}
```

You would need custom serializers to make GSON work as is.  Jackson just works.

**Comment by bbakerman** [comment]
BTW @andimarek - re This would also solve this PR: #647 . Custom data can be added after toSpecification.

It might not because thatb PR is asking for  a way to "transfer" custom data from an exception in a data fetcher into the actual ErrorWhileDataFetching class.

That is make it have user defined values but surface them up into a number of errors

**Comment by bbakerman** [comment]
More testing has revealed some problems

     public class NonNullableFieldWasNullException extends RuntimeException implements GraphQLError {

        private final ExecutionTypeInfo typeInfo;
         private final ExecutionPath path;

This wont serialize in neither Jackson or GSON.  Both fail because the graphql types (in ExecutionTypeInfo I think)  are circular somehow and it both get a stack overflow exception.

The NonNullableFieldWasNullException as is a problematic error to report

**Comment by andimarek** [comment]
But more reasons to provide a specific method to convert it something which can easily be converted by every json library, or do I miss something?

**Comment by bbakerman** [comment]
I have been looking more closely at GraphqlError and I am wondering the value of ErrorType getErrorType();
 public enum ErrorType {

    InvalidSyntax,
    ValidationError,
    DataFetchingException,
    MutationNotSupported

}

its not spec - but it does classify errors - but only kinda

**Comment by andimarek** [comment]
yes, my initial design idea was to classify the errors without creating specific subclasses. But I am not sure how up to date this list is and I am not sure how useful it is overall.

**Comment by andimarek** [comment]
this will be solved by #650
